nodejs/awsx/ec2/vpcTopology.ts

```diff
@@ -18,42 +18,243 @@ import * as pulumi from "@pulumi/pulumi";
 import * as x from "..";
 import { Cidr32Block, getIPv4Address } from "./cidr";
 
+import * as utils from "./../utils";
+
+/** @internal */
+export interface AvailabilityZoneDescription {
+    name: string;
+    id: string;
+}
+
 /** @internal */
 export class VpcTopology {
     private readonly vpcCidrBlock: Cidr32Block;
     private lastAllocatedSubnetCidrBlock?: Cidr32Block;
 
-    constructor(private readonly vpcName: string,
+    constructor(private readonly resource: pulumi.Resource | undefined,
+                private readonly vpcName: string,
                 vpcCidr: string,
-                private readonly numberOfAvailabilityZones: number) {
+                private readonly ipv6CidrBlock: pulumi.Output<string> | undefined,
+                private readonly availabilityZones: AvailabilityZoneDescription[],
+                private readonly numberOfNatGateways: number,
+                private readonly assignGeneratedIpv6CidrBlock: pulumi.Input<boolean>) {
 
         this.vpcCidrBlock = Cidr32Block.fromCidrNotation(vpcCidr);
     }
 
-    public createSubnets(subnetArgsArray: x.ec2.VpcSubnetArgs[]) {
+    public create(subnetArgsArray: x.ec2.VpcSubnetArgs[]): VpcTopologyDescription {
+        // Check if the subnets were given explicit location information or not.  If so, we'll
+        // respect the location information the user asked for.  If not, we'll automatically
+        // smartly partition the vpc.
+        const firstSubnet = subnetArgsArray[0];
+        const hasExplicitLocation = !!firstSubnet.location;
+        for (const subnet of subnetArgsArray) {
+            const siblingHasLocation = !!subnet.location;
+
+            if (hasExplicitLocation !== siblingHasLocation) {
+                throw new pulumi.ResourceError("[location] property must be specified for either no subnets or all of the subnets.", this.resource);
+            }
+
+            if (siblingHasLocation && subnet.cidrMask !== undefined) {
+                throw new pulumi.ResourceError("Subnet cannot specify [location] and [cidrMask]", this.resource);
+            }
+        }
```

Comment:
these above checks were new.  to ensure all the subnet descriptions were consistent.